plenum/server/consensus/checkpoint_service.py

```diff
@@ -368,58 +234,51 @@ def update_watermark_from_3pc(self):
-                del self._stashed_recvd_checkpoints[view_no]
+        def is_below_3pc_key(cp: CheckpointService.CheckpointKey, key: Tuple[int, int]):
+            if cp.view_no < key[0]:
+                return True
+            if cp.view_no > key[0]:
+                return False
+            return cp.pp_seq_no <= till_3pc_key[1]
 
-            elif view_no == till_3pc_key[0]:
-                for (s, e) in list(self._stashed_recvd_checkpoints[view_no].keys()):
-                    if e <= till_3pc_key[1]:
-                        self._logger.info('{} removing stashed checkpoints: '
-                                          'viewNo={}, seqNoStart={}, seqNoEnd={}'.
-                                          format(self, view_no, s, e))
-                        del self._stashed_recvd_checkpoints[view_no][(s, e)]
-                if len(self._stashed_recvd_checkpoints[view_no]) == 0:
-                    del self._stashed_recvd_checkpoints[view_no]
+        for cp in list(self._received_checkpoints.keys()):
+            if is_below_3pc_key(cp, till_3pc_key):
+                self._logger.info('{} removing received checkpoints: {}'.format(self, cp))
+                del self._received_checkpoints[cp]
 
     def _reset_checkpoints(self):
         # That function most probably redundant in PBFT approach,
         # because according to paper, checkpoints cleared only when next stabilized.
         # Avoid using it while implement other services.
-        self._checkpoint_state.clear()
         self._data.checkpoints.clear()
         # TODO: change to = 1 in ViewChangeService integration.
         self._data.stable_checkpoint = 0
 
-    def _set_stable_checkpoint(self, end_seq_no):
-        if not list(self._data.checkpoints.irange_key(end_seq_no, end_seq_no)):
-            raise LogicError('Stable checkpoint must be in checkpoints')
-        self._data.stable_checkpoint = end_seq_no
-
-        self._data.checkpoints = \
-            SortedListWithKey([c for c in self._data.checkpoints if c.seqNoEnd >= end_seq_no],
-                              key=lambda checkpoint: checkpoint.seqNoEnd)
-
     def __str__(self) -> str:
         return "{} - checkpoint_service".format(self._data.name)
 
-    # TODO: move to OrderingService as a handler for Cleanup messages
-    # def _clear_batch_till_seq_no(self, seq_no):
-    #     self._data.preprepared = [pp for pp in self._data.preprepared if pp.ppSeqNo >= seq_no]
-    #     self._data.prepared = [p for p in self._data.prepared if p.ppSeqNo >= seq_no]
-
     def discard(self, msg, reason, sender):
         self._logger.trace("{} discard message {} from {} "
                            "with the reason: {}".format(self, msg, sender, reason))
+
+    def _have_own_checkpoint(self, key: CheckpointKey) -> bool:
+        own_checkpoints = self._data.checkpoints.irange_key(min_key=key.pp_seq_no, max_key=key.pp_seq_no)
```

Comment:
How can we have multiple Checkpoints for the same ppSeqNo?

Reply:
Logically this case should be impossible, but data structure (sorted list) allows it. Probably dictionary may turn out to be better fit, will consider it in next PR.